component-catalog-connectors/artifactory-connector/artifactory_catalog_connector/artifactory_catalog_connector.py

```diff
@@ -0,0 +1,403 @@
+                                 stored; in addition to catalog_entry_data, catalog_metadata may also be
+                                 needed to read the component definition for certain types of catalogs
+        :returns: an EntryData object representing the definition (and other identifying info) for a single
+                  catalog entry; if None is returned, this catalog entry is skipped and a warning message logged
+        """
+        ########################
+        # required metadata
+        ########################
+        entry_url = catalog_entry_data.get("url")
+        if entry_url is None:
+            self.log.error("Artifactory component entries must specify `url`")
+            return None
+
+        ########################
+        # optional metadata
+        ########################
+        artifactory_username = catalog_metadata.get("artifactory_username")
+        artifactory_password = catalog_metadata.get("artifactory_password")
+
+        # parse `artifactory_username` and `artifactory_password`
+        api_auth = None
+        if artifactory_username and artifactory_password:
+            api_auth = HTTPBasicAuth(
+                username=artifactory_username, password=artifactory_password
+            )
+        elif artifactory_password:
+            api_auth = ArtifactoryApiKeyAuth(api_key=artifactory_password)
+
+        ########################
+        # get component
+        ########################
+        try:
+            self.log.debug(
+                f"Retrieving component from Artifactory catalog '{entry_url}'"
+            )
+            resp = requests.get(entry_url, auth=api_auth)
+
+            # verify HTTP status code
+            if resp.status_code != HTTPStatus.OK:
+                raise RuntimeError(
+                    f"Got unhandled HTTP Status {resp.status_code}, expected 200... "
+                    f"{resp.text}"
+                )
+
+            # verify Content-Type
+            resp_content_type = resp.headers.get("Content-Type", "")
+            expected_content_type = "text/plain"
+            if resp_content_type != expected_content_type:
+                raise RuntimeError(
+                    f"Got unexpected Content-Type '{resp_content_type}', "
+                    f"expected '{expected_content_type}'"
+                )
+        except Exception as ex:
+            self.log.error(
+                f"Error retrieving component specification from '{entry_url}': {ex}"
+            )
+            return None
+
+        return KfpEntryData(definition=resp.text)
+
+    @classmethod
+    def get_hash_keys(cls) -> List[Any]:
+        """
+        Identifies the unique Artifactory catalog key that get_entry_data
+        can use to fetch a component from the catalog. Method get_catalog_entries
+        retrieves the list of available key values from the catalog.
+
+        :returns: a list of keys
+        """
+        return ["url"]
```

Comment:
@ptitzler while I know it makes things a little less portable, I think it's safer to use the full URL as the hash ([like the URL Connector](https://github.com/elyra-ai/elyra/blob/001b4ff28bdfb2751d4649047dc50a1af4f01016/elyra/pipeline/catalog_connector.py#L686)), this is because:

1. The URL may-or-may-not be under the same host/port/path as the one we initiated the search from, so I am not even sure how to extract a sensible repository_name / file_path combo.
2. Having the full URL in the component description, lets people more easily browse to that URL and download it for themselves (when looking at the "component source" JSON). 
3. In my experience, people never change the host/port of their Artifactory server, as it would break heaps of still (like this), so I am not sure it will affect many people.

Reply:
Okay, fair enough. We can always revisit once there are others using the connector!